### PR TITLE
Why Lucky?

### PR DESCRIPTION
**Pull request by paulcsmith**
Preview of the page: https://deploy-preview-57--registrar-cat-46731.netlify.com/why_lucky/

Closes #47 

This is not an exhaustive list and I'd like to add more/better examples at some point, but I'm hoping this is at least a good start for why lucky exists and why people might want to use it.

### CODE REVIEW

File: source/why_lucky/index.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+layout: with-intro
+title: Why Lucky?
+intro: >
+  Catch bugs early, forget about most performance issues,
+  and spend more time on code instead of debugging and writing tests.
+---
+
+## The Lucky philosophy
+
+Lucky was designed to solve a few core problems that teams often see. Lucky strives to:
```
Comment (paulcsmith) [inline source/why_lucky/index.html.md.erb:11]: Here is where the new page has been added. The rest is just setting up some different HTML layouts to accommodate this new page

File: source/why_lucky/index.html.md.erb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  than everyone else.
+
+Lucky was designed for developers that love making fast and reliable products.
+We think you'll love it.
+
+## Spend less time writing tests and debugging
+
+Type safe database queries, rock solid routing, type safe forms and validations,
+and more. This is how Lucky helps you find errors before they reach your
+customers, write fewer tests, and spend less time fixing emberassing bugs.
```
Comment (ugisozols) [inline source/why_lucky/index.html.md.erb]: `emberassing` -> `embarrassing`.

File: source/why_lucky/index.html.md.erb [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+# Even complex HTML and JSON renders in just a couple milliseconds
+# This is rendering a simple HTML page in 1/10th of a millisecond
+Rendered GET /tasks - 200 (94.0µs)
+```
+
+## Strong type safety with minimal boilerplate
+
+Lucky generates type safe and reliable code for you, and makes things type safe
+that many frameworks can't or don't. This means you spend less time shipping
+emberassing bugs to customers.
```
Comment (ugisozols) [inline source/why_lucky/index.html.md.erb]: `emberassing` -> `embarrassing`.

File: source/why_lucky/index.html.md.erb [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+assets and will even suggest the right one if you have a typo.
+
+```plain
+"images/logo.jpeg" does not exist in the manifest
+Did you mean "images/logo.jpg"?
+```
+
+## Reliable and future proof configuration
+
+Lucky will fail to to start if you forget a required configuration option, and
```
Comment (ugisozols) [inline source/why_lucky/index.html.md.erb]: 2x `to`.

File: source/why_lucky/index.html.md.erb [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+  LuckyWeb::Session::Store.configure do
+    settings.secret = "some_value"
+  end
+```
+
+## Update dependencies with confidence
+
+Updating your dependencies will always be difficult, but Crystal makes it a bit
+easier. If a method name changes, or it’s arguments are different from the
```
Comment (ugisozols) [inline source/why_lucky/index.html.md.erb]: `it's` -> `its`.

File: source/why_lucky/index.html.md.erb (hunk 2 above)
Comment (paulcsmith) [reply source/why_lucky/index.html.md.erb]: That would have been embarrassing :) haha

File: source/why_lucky/index.html.md.erb [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  than everyone else.
+
+Lucky was designed for developers that love making fast and reliable products.
+We think you'll love it.
+
+## Spend less time writing tests and debugging
+
+Type safe database queries, rock solid routing, type safe forms and validations,
+and more. This is how Lucky helps you find errors before they reach your
+customers, write fewer tests, and spend less time fixing embarassing bugs.
```
Comment (dbernheisel) [inline source/why_lucky/index.html.md.erb]: You give some lightweight examples for most other reasons, but I don't see one here.

File: source/why_lucky/index.html.md.erb [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+Lucky generates type safe and reliable code for you, and makes things type safe
+that many frameworks can't or don't. This means you spend less time shipping
+embarassing bugs to customers.
+
+For example, when you define a model it will create some classes and methods for
+querying, creating, and updating records in a type safe way. If you write a
+query with a mistyped column name or with the wrong type, Crystal will let you
+know. If you try to use a field that doesn’t exist in an HTML form, it'll let
+you know.
```
Comment (dbernheisel) [inline source/why_lucky/index.html.md.erb]: You give some lightweight examples for most other reasons, but I don't see one here.

**Comment by dbernheisel** [comment]
I left a comment about maybe adding another example or two. Otherwise, this looks great 👍

File: source/why_lucky/index.html.md.erb [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+UserQuery.new.age.gte(30).admin(true).email.ilike("%thoughtbot.com")
+```
+
+## Send content to users _fast_
+
+Lucky helps you spend less time fixing performance issues, and more time
+delighting your customers with fast applications.
+
+```
+# Even complex HTML and JSON renders in just a couple milliseconds
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: It'd be cool to say how long it took to render such a page and link to the code, but I get that there's not really an appropriate demo to link to. Maybe a "lines of outputted HTML"? I'm sure one day we'll write the guides in Lucky 🙂 

I think I'm feeling that I want a little better idea of how significant this claim is. _How_ fast? _How_ complex?

File: source/why_lucky/index.html.md.erb [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+Rendered GET /tasks - 200 (94.0µs)
+```
+
+## Strong type safety with minimal boilerplate
+
+Lucky generates type safe and reliable code for you, and makes things type safe
+that many frameworks can't or don't. This means you spend less time shipping
+embarassing bugs to customers.
+
+For example, when you define a model it will create some classes and methods for
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: Maybe link `define a model` to the _Querying the Database_ guide?

File: source/why_lucky/index.html.md.erb [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+querying, creating, and updating records in a type safe way. If you write a
+query with a mistyped column name or with the wrong type, Crystal will let you
+know. If you try to use a field that doesn’t exist in an HTML form, it'll let
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: I think "field that doesn't exist" is kind of confusing? I'm not sure what you mean. The first thought I had was "how would I try to use it if it doesn't exist?" Maybe it's a bit pedantic but I can't parse it.

Maybe you meant the attribute doesn't exist on the model? If so, I'd rephrase this to something like:

> If you try to create a form field for a non-existent attribute, it'll let you know.

File: source/why_lucky/index.html.md.erb [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+you know.
+
+## Avoid `nil` errors
+
+Instead of nil errors in production, Crystal and Lucky help you know about nil
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: This reads a little awkward, I think. Maybe:

> ... Crystal and Lucky tell you about nill errors at compile time ...

File: source/why_lucky/index.html.md.erb [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+errors at compile time, before your customers ever see them. Lucky has designed
+its router, HTML, actions, params, and forms so that Crystal can catch as many
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: It would be great to link router (actions and routing), HTML (rendering), and forms to the relevant guides. People are probably going to read this before others (if they read it at all) so letting them know there is more relevant material to read might draw them in.

Comment (edwardloveall) [reply source/why_lucky/index.html.md.erb]: Maybe params too? I don't remember if there's a guide or section that covers params explicitly.

File: source/why_lucky/index.html.md.erb [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+It will also fail to compile if one of your dependencies changes its
+configuration options, and let you know what to do to fix it.
+
+```
+LuckyWeb::Session::Store.configure do
+  settings.key = "my_app"
+end
+
+# When you try to compile, you'll see this:
+LuckyWeb::Session::Store.settings.secret was nil, but the setting is required. Please set it.
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: Are all settings required implicitly? I'm not sure I understand how to mark a setting as required.

File: source/why_lucky/index.html.md.erb [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+automatically recompiled and the page is refreshed. On top of that,
+[Browsersync](https://www.browsersync.io) lets you connect through a proxy to
+easily test pages on your mobile device, or simulate slow connections to make
+sure all your customers are happy, even on slow connections.
+
+## Never let an unhandled form param through
+
+Lucky makes sure only allowed parameters are saved to the database. It even
+catches issues at compile time when you try to add a form input for a parameter
+that isn’t allowed.
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb:161]: Is this related to my question up above? If so, it makes sense down here.

File: source/why_lucky/index.html.md.erb [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+  and spend more time on code instead of debugging and writing tests.
+---
+
+## The Lucky philosophy
+
+Lucky was designed to solve a few core problems that teams often see. Lucky strives to:
+
+* Catch bugs at compile time, rather than finding them in production.
+* Spend less time writing tests, because the compiler catches many errors for you.
+* Minimize guess work by using conventions for the most common tasks.
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: I believe [guesswork](http://www.wordnik.com/words/guesswork) is one word.

File: source/why_lucky/index.html.md.erb [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+* Help developers break things into discrete pieces so things are easy to share
+  and maintain in the future.
+* Minimize boiler plate code so it's easy to focus on what your app does better
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: [Boilerplate](https://en.wikipedia.org/wiki/Boilerplate_text) is one word

File: source/why_lucky/index.html.md.erb (hunk 6 above)
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: ~embarassing~ embarrassing 😉

File: source/why_lucky/index.html.md.erb [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+Type safe database queries, rock solid routing, type safe forms and validations,
+and more. This is how Lucky helps you find errors before they reach your
+customers, write fewer tests, and spend less time fixing embarassing bugs.
+
+## Type safe database queries
+
+Lucky will help catch bad queries at compile time. Everything is chainable and
+the methods you can use are determined by the type. So strings can use `ilike`
+and `lower`, but if you try to use those on a integer column, Lucky will tell
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: _an_ integer

File: source/why_lucky/index.html.md.erb [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+```
+# Even complex HTML and JSON renders in just a couple milliseconds
+# This is rendering a simple HTML page in 1/10th of a millisecond
+Rendered GET /tasks - 200 (94.0µs)
+```
+
+## Strong type safety with minimal boilerplate
+
+Lucky generates type safe and reliable code for you, and makes things type safe
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: No need for the comma here

File: source/why_lucky/index.html.md.erb [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+that many frameworks can't or don't. This means you spend less time shipping
+embarassing bugs to customers.
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: ~embarassing~ embarrassing

File: source/why_lucky/index.html.md.erb (hunk 11 above)
Comment (edwardloveall) [reply source/why_lucky/index.html.md.erb]: Also no need for the comma after "compile time"

File: source/why_lucky/index.html.md.erb [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+assets and will even suggest the right one if you have a typo.
+
+```plain
+"images/logo.jpeg" does not exist in the manifest
+Did you mean "images/logo.jpg"?
+```
+
+## Reliable and future proof configuration
+
+Lucky will fail to start if you forget a required configuration option, and
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: No need for the comma after option

File: source/why_lucky/index.html.md.erb [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+
+  LuckyWeb::Session::Store.configure do
+    settings.secret = "some_value"
+  end
+```
+
+## Update dependencies with confidence
+
+Updating your dependencies will always be difficult, but Crystal makes it a bit
+easier. If a method name changes, or its arguments are different from the
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: No need for the comma after "changes"

File: source/why_lucky/index.html.md.erb [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+Never accidentally print `nil` to the page, extract and share partials using
+regular methods. Easily read an entire page by looking at just the render
+method. Text is automatically escaped for security. And it's all type safe. That
+means no more unmatched closing tags, and never rendering a page with missing data.
+
+## Fault proof HTTP verbs
+
+REST can be a bit unwieldy at times. It's easy to use the right path and wrong
+verb and then wonder why the router can't find a matching route. Lucky makes
+this easy by setting both the path and HTTP method when generating links, forms
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: Oxford comma after forms

File: source/why_lucky/index.html.md.erb [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+and redirects.
+
+```
+# Lucky will set the path and the DELETE HTTP method
+link "Delete", to: Tasks::Delete.with(id: task.id)
+```
+
+## Built in live reloading with Browsersync
+
+CSS and JS reloads almost instantly, and every change to source code is
```
Comment (edwardloveall) [inline source/why_lucky/index.html.md.erb]: Singular "reload"

**Review by edwardloveall — COMMENTED** [review]
Great points here. Just a couple of comments about wording, clarity, and grammar

File: source/why_lucky/index.html.md.erb (hunk 13 above)
Comment (paulcsmith) [reply source/why_lucky/index.html.md.erb]: They are marked as required/not required by the module/class. So `LuckyWeb::Session:Store` declares that the key is required. http://luckyframework.org/guides/configuration/#configuring-your-own-code

File: source/why_lucky/index.html.md.erb (hunk 14 above)
Comment (paulcsmith) [reply source/why_lucky/index.html.md.erb:161]: Yeah it is. I'll try to rework it a bit :)

File: source/why_lucky/index.html.md.erb (hunk 6 above)
Comment (paulcsmith) [reply source/why_lucky/index.html.md.erb]: This is a more general thing so I'm not sure I have one example. If I think of something I will add one though!

**Review by paulcsmith — COMMENTED** [review]
Thanks for all the feedback! I'll make some changes and merge this in. I'll try to add even more examples in a future update too

File: source/why_lucky/index.html.md.erb (hunk 8 above)
Comment (paulcsmith) [reply source/why_lucky/index.html.md.erb]: I do agree a better benchmark, with a link and some req/s would be good. I don't have time for that quite yet, but I will definitely do this. For now I added a slightly more concrete description.
